spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -396,45 +396,57 @@ public void close() throws IOException {
       // schema or rows returned by readers
       Schema finalSchema = expectedSchema;
       PartitionSpec spec = task.spec();
-      Set<Integer> idColumns = spec.identitySourceIds();
+      Set<Integer> idColumns = Sets.newHashSet();
+      for (Integer i : spec.identitySourceIds()) {
+        if (spec.schema().columns().stream()
+            .noneMatch(j -> j.type().isStructType() && j.type().asStructType().field(i) != null)
+        ) {
+          idColumns.add(i);
+        }
+      }
 
-      // schema needed for the projection and filtering
       StructType sparkType = SparkSchemaUtil.convert(finalSchema);
       Schema requiredSchema = SparkSchemaUtil.prune(tableSchema, sparkType, task.residual(), caseSensitive);
       boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
       boolean hasExtraFilterColumns = requiredSchema.columns().size() != finalSchema.columns().size();
 
-      Schema iterSchema;
       Iterator<InternalRow> iter;
 
       if (hasJoinedPartitionColumns) {
-        // schema used to read data files
-        Schema readSchema = TypeUtil.selectNot(requiredSchema, idColumns);
         Schema partitionSchema = TypeUtil.select(requiredSchema, idColumns);
         PartitionRowConverter convertToRow = new PartitionRowConverter(partitionSchema, spec);
-        JoinedRow joined = new JoinedRow();
 
         InternalRow partition = convertToRow.apply(file.partition());
-        joined.withRight(partition);
 
-        // create joined rows and project from the joined schema to the final schema
-        iterSchema = TypeUtil.join(readSchema, partitionSchema);
-        iter = Iterators.transform(open(task, readSchema), joined::withLeft);
+        iter = Iterators.transform(open(task, finalSchema), input -> {
+          InternalRow internalRow;
+          if (input == null) {
+            internalRow = new GenericInternalRow(spec.fields().size());
+          } else {
+            internalRow = input;
+          }
+          int index = 0;
+          for (PartitionField pf : spec.fields()) {
+            int ref = pf.sourceId() - 1;
+            if (index < partition.numFields()) {
+              if (ref > internalRow.numFields()) {
+                internalRow.setNullAt(ref);
+              }
+              internalRow.update(ref, ((GenericInternalRow) partition).genericGet(index));
```

Comment:
Looks like what you're trying to do is match up the partition fields with fields in the schema, using the field index for data coming from partition and the source ID to recover an index into the output row.

Unfortunately, there are several problems with this approach. The biggest one is that the source ID can't be used to produce an index. This may appear to work when all source fields are projected, but if you don't project a field, then that index is off by 1. It also assumes that IDs are assigned a certain way and there is not guarantee for that in the format. Reordering fields also causes problems with this approach.

The problems can be fixed by using the approach I recommended earlier: this should build a map of source ID to partition value. Then it should pass that map in when building the [reader tree](https://github.com/apache/incubator-iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/data/SparkParquetReaders.java#L70), so that the tree can use constant readers instead of primitive readers. Here's how it happens in Pig: https://github.com/apache/incubator-iceberg/blob/master/pig/src/main/java/org/apache/iceberg/pig/PigParquetReader.java#L156-L160